cmake/Modules/FindGCSSDK_EP.cmake

```diff
@@ -0,0 +1,165 @@
+    ${TILEDB_DEPS_NO_DEFAULT_PATH}
+  )
+  find_path(GCSSDK_INCLUDE_DIR
+    NAMES client.h
+    PATH_SUFFIXES include/google/cloud/storage
+    ${TILEDB_DEPS_NO_DEFAULT_PATH}
+  )
+endif()
+
+include(FindPackageHandleStandardArgs)
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(GCSSDK
+  REQUIRED_VARS GCSSDK_LIBRARY GCSSDK_INCLUDE_DIR
+)
+
+if (NOT GCSSDK_FOUND)
+  if (TILEDB_SUPERBUILD)
+    message(STATUS "Could NOT find GCSSDK")
+    message(STATUS "Adding GCSSDK as an external project")
+
+    set(DEPENDS)
+    if (TARGET ep_curl)
+      list(APPEND DEPENDS ep_curl)
+    endif()
+    if (TARGET ep_openssl)
+      list(APPEND DEPENDS ep_openssl)
+    endif()
+    if (TARGET ep_zlib)
+      list(APPEND DEPENDS ep_zlib)
+    endif()
+
+    # Fetch the number of CPUs on the this sytem.
+    include(ProcessorCount)
+    processorcount(NCPU)
+
+    ExternalProject_Add(ep_gcssdk
+      PREFIX "externals"
+      URL "https://github.com/googleapis/google-cloud-cpp/archive/v0.20.0.zip"
+      URL_HASH SHA1=6e7931a0e62779dfbd07427373373bf1ad6f8686
+      BUILD_IN_SOURCE 1
+      PATCH_COMMAND
+        patch -N -p1 < ${TILEDB_CMAKE_INPUTS_DIR}/patches/ep_gcssdk/build.patch &&
+        patch -N -p1 < ${TILEDB_CMAKE_INPUTS_DIR}/patches/ep_gcssdk/ls.patch
+      CONFIGURE_COMMAND
+        cmake -Hsuper -Bcmake-out
+        -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+        -DBUILD_SHARED_LIBS=OFF
+        -DBUILD_SAMPLES=OFF
+        -DCMAKE_PREFIX_PATH=${TILEDB_EP_INSTALL_PREFIX}
+        -DOPENSSL_ROOT_DIR=${TILEDB_OPENSSL_DIR}
+      BUILD_COMMAND cmake --build cmake-out -- -j${NCPU}
+      INSTALL_COMMAND
+          cp -r ${TILEDB_EP_SOURCE_DIR}/ep_gcssdk/cmake-out/external/lib ${TILEDB_EP_INSTALL_PREFIX}
+        COMMAND
+          cp -r ${TILEDB_EP_SOURCE_DIR}/ep_gcssdk/cmake-out/external/include ${TILEDB_EP_INSTALL_PREFIX}
+      LOG_DOWNLOAD TRUE
+      LOG_CONFIGURE TRUE
+      LOG_BUILD TRUE
+      LOG_INSTALL TRUE
+      LOG_OUTPUT_ON_FAILURE ${TILEDB_LOG_OUTPUT_ON_FAILURE}
+      DEPENDS ${DEPENDS}
+    )
+
+    list(APPEND TILEDB_EXTERNAL_PROJECTS ep_gcssdk)
+    list(APPEND FORWARD_EP_CMAKE_ARGS
+      -DTILEDB_GCSSDK_EP_BUILT=TRUE
+    )
+
+  endif()
+endif()
+
+if (GCSSDK_FOUND)
+  # Build a list of all GCS libraries to link with.
+  list(APPEND GCSSDK_LINKED_LIBS "storage_client"
+                                 "google_cloud_cpp_common"
+                                 "crc32c")
+
+  foreach (LIB ${GCSSDK_LINKED_LIBS})
+    find_library(GCS_FOUND_${LIB}
+      NAMES lib${LIB}${CMAKE_STATIC_LIBRARY_SUFFIX}
+      PATHS ${TILEDB_EP_INSTALL_PREFIX}
+      PATH_SUFFIXES lib
+      ${TILEDB_DEPS_NO_DEFAULT_PATH}
+    )
+
+    message(STATUS "Found GCS lib: ${LIB} (${GCS_FOUND_${LIB}})")
+    if (NOT TARGET GCSSDK::${LIB})
+      add_library(GCSSDK::${LIB} UNKNOWN IMPORTED)
+      set_target_properties(GCSSDK::${LIB} PROPERTIES
+        IMPORTED_LOCATION "${GCS_FOUND_${LIB}}"
+        INTERFACE_INCLUDE_DIRECTORIES ${TILEDB_EP_INSTALL_PREFIX}/include
```

Comment:
@ihnorton This line fixes the issue I was describing on Friday when I said that the the OSX CI build was failing to find the GCS headers. Note that I said this was specific to static-builds but that was an error in my testing, the issue was for both shared and static builds.

Let's consider the scenario where the build failed on `#include <google/cloud/storage/client.h>`:
- The file exists at `build/externals/install/include/google/cloud/storage/client.h`
- This line was previously setting `INTERFACE_INCLUDE_DIRECTORIES` to ${GCSSDK_INCLUDE_DIR}. This would expand to `build/externals/install/include/google/cloud/storage`.
- The include statement would fail looking for `build/externals/install/include/google/cloud/storage/google/cloud/storage/client.h`.

This issue was hidden on local builds because other external projects were placing their headers directly in `build/externals/install/include` so the search path included `build/externals/install/include`. On CI, this was not happening because it found the headers for the other external project elsewhere.

While debugging this, I noticed that many of the include paths were in the source directories rather than the install directory (`build/externals/install/include`), for example:
```
-I/Users/runner/runners/2.165.2/work/1/s/build/externals/src/ep_tbb/include
-I/Users/runner/runners/2.165.2/work/1/s/build/externals/src/ep_spdlog/include
```

I think the long-term fix is to just just include `build/externals/install/include` and 1) ensure all headers are installed properly into that directory, 2) ensure that directory always exists regardless of whether we have any external projects, and 3) remove logic like this where we place the include directories in the external target libraries.

Reply:
I agree about making a clean separation between the `ep_build` directories and the install directory.

I believe the issue described above is that the include signature for `find_path(GCSSDK_INCLUDE_DIR ...)` needs to match the `#include` usage, eg:

```
  find_path(GCSSDK_INCLUDE_DIR
    NAMES client.h
    PATHS ${TILEDB_EP_INSTALL_PREFIX}
    PATH_SUFFIXES include/google/cloud/storage
    NO_DEFAULT_PATH
  )
```

should be

```
    NAMES "google/cloud/storage/client.h
    PATHS ${TILEDB_EP_INSTALL_PREFIX}/include
    ...
```

otherwise you will get back the fully-prefixed path. But I can only imagine how this was hard to debug, as it seems that `find_path` ignores `unset(VARNAME)` (and just uses the cached value).